bukuserver/api.py

```diff
@@ -110,178 +148,157 @@ def put(self, tag: str):
-            all_bookmarks = bukudb.get_rec_all()
-            result = {'bookmarks': [entity(bookmark, id=not request.path.startswith('/api/'))
-                                    for bookmark in all_bookmarks]}
-        else:
-            bukudb = getattr(flask.g, 'bukudb', get_bukudb())
-            bookmark = bukudb.get_rec_by_id(rec_id)
-            if bookmark is None:
-                return Response.BOOKMARK_NOT_FOUND()
-            result = entity(bookmark)
-        return Response.SUCCESS(data=result)
+class ApiBookmarksView(MethodView):
+    def get(self):
+        bukudb = getattr(flask.g, 'bukudb', get_bukudb())
+        order = request.args.getlist('order')
+        all_bookmarks = bukudb.get_rec_all(order=order)
+        return Response.SUCCESS(data={'bookmarks': [entity(bookmark, index=order)
+                                                    for bookmark in all_bookmarks]})
 
-    def post(self, rec_id: None = None):
-        form = ApiBookmarkCreateForm({})
-        error_response, error_data = form.process_data(request.get_json())
-        if error_response is not None:
-            return error_response(data=error_data)
+    def post(self):
+        try:
+            form = ApiBookmarkCreateForm(data=request.get_json())
+        except BadRequest:
+            return Response.INVALID_REQUEST()
+        if not form.validate():
+            return Response.invalid(form.errors)
         bukudb = getattr(flask.g, 'bukudb', get_bukudb())
-        result_flag = bukudb.add_rec(
+        index = bukudb.add_rec(
             form.url.data,
             form.title.data,
             form.tags_str,
             form.description.data,
             fetch=form.fetch.data)
-        return Response.from_flag(result_flag)
+        return Response.from_flag(index is not None, data=index and {'index': index})
 
-    def put(self, rec_id: int):
-        form = ApiBookmarkEditForm({})
-        error_response, error_data = form.process_data(request.get_json())
-        if error_response is not None:
-            return error_response(data=error_data)
+    def delete(self):
         bukudb = getattr(flask.g, 'bukudb', get_bukudb())
-        result_flag = bukudb.update_rec(
-            rec_id,
-            form.url.data,
-            form.title.data,
-            form.tags_str,
-            form.description.data)
-        return Response.from_flag(result_flag)
+        return Response.from_flag(bukudb.cleardb(confirm=False))
 
-    def delete(self, rec_id: T.Union[int, None]):
-        if rec_id is None:
-            bukudb = getattr(flask.g, 'bukudb', get_bukudb())
-            with mock.patch('buku.read_in', return_value='y'):
-                result_flag = bukudb.cleardb()
-        else:
-            bukudb = getattr(flask.g, 'bukudb', get_bukudb())
-            result_flag = bukudb.delete_rec(rec_id, retain_order=True)
-        return Response.from_flag(result_flag)
 
+class ApiBookmarkView(MethodView):
+    def get(self, index: int):
```

Comment:
Incidentally, all occurrences of "IDs" got renamed into "indices" – since it more accurately matches how such values are used in the app